src/main/java/slimeknights/tconstruct/tables/tileentity/table/tinkerstation/TinkerStationTileEntity.java

```diff
@@ -0,0 +1,215 @@
+    return new TinkerStationContainer(menuId, playerInventory, this);
+  }
+
+  /* Crafting */
+
+  @Override
+  public ItemStack calcResult() {
+    if (this.world == null) {
+      return ItemStack.EMPTY;
+    }
+
+    // assume empty unless we learn otherwise
+    ItemStack result = ItemStack.EMPTY;
+
+    if (!this.world.isRemote && this.world.getServer() != null) {
+      RecipeManager manager = this.world.getServer().getRecipeManager();
+
+      // first, try the cached recipe
+      ITinkerStationRecipe recipe = lastRecipe;
+      // if it does not match, find a new recipe
+      if (recipe == null || !recipe.matches(this.inventoryWrapper, this.world)) {
+        recipe = manager.getRecipe(RecipeTypes.TINKER_STATION, this.inventoryWrapper, this.world).orElse(null);
+      }
+
+      // if we have a recipe, fetch its result
+      if (recipe != null) {
+        result = recipe.getCraftingResult(this.inventoryWrapper);
+        // sync if the recipe is different
+        if (recipe != this.lastRecipe) {
+          this.lastRecipe = recipe;
+          this.syncToRelevantPlayers();
+        }
+      }
+    }
+    else if (this.lastRecipe != null && this.lastRecipe.matches(this.inventoryWrapper, world)) {
+      result = this.lastRecipe.getCraftingResult(this.inventoryWrapper);
+    }
+
+    return result;
+  }
+
+  @Override
+  public ItemStack onCraft(PlayerEntity player, ItemStack result, int amount) {
+    if (this.world == null || amount == 0 || this.lastRecipe == null || !this.lastRecipe.matches(this.inventoryWrapper, world)) {
+      return ItemStack.EMPTY;
+    }
+
+    // check if the player has access to the result
+    if (player instanceof ServerPlayerEntity) {
+      if (this.lastRecipe != null) {
+        // if the player cannot craft this, block crafting
+        if (!this.lastRecipe.isDynamic() && world.getGameRules().getBoolean(GameRules.DO_LIMITED_CRAFTING) && !((ServerPlayerEntity) player).getRecipeBook().isUnlocked(this.lastRecipe)) {
+          return ItemStack.EMPTY;
+        }
+        // unlock the recipe if it was not unlocked
+        if (this.lastRecipe != null && !this.lastRecipe.isDynamic()) {
+          player.unlockRecipes(Collections.singleton(this.lastRecipe));
+        }
+      }
+
+      // fire crafting events
+      result.onCrafting(this.world, player, amount);
+      BasicEventHooks.firePlayerCraftingEvent(player, result, this.inventoryWrapper);
+    }
+
+    this.playCraftSound(player);
+
+    // update all slots in the inventory
+    // remove remaining items
+    NonNullList<ItemStack> remaining = this.lastRecipe.getRemainingItems(this.inventoryWrapper);
+
+    for (int i = 0; i < remaining.size(); ++i) {
+      ItemStack original = getStackInSlot(i);
+      ItemStack newStack = remaining.get(i);
+
+      // if the slot contains a stack, decrease by 1
+      if (!original.isEmpty()) {
+        original.shrink(1);
```

Comment:
Going to need a way for the recipe to specify a higher number eventually

Reply:
Yes, I think we should do that once we start working on the repair recipe and such

Reply:
One idea is making a method in `ITinkerStationRecipe`:

```java
default int getAmountUsed(int index) {
  return 1;
}
```
We then call the method here by incrementing the index only when we see a non empty slot that is not the tool slot.